### ISSUE TITLE
Support version letter in shlib version number

### ISSUE THREAD

**Issue by dvarrazzo** [openssl/openssl#2880]
If `SHLIB_VERSION_NUMBER` is set to a value including the version letter, e.g. `1.0.2k`, I think the `.so` name should respect that and include the letter. Currently it generates a library stopping to the minor release, i.e. `1.0.2`.

Proposed change:

```diff
diff --git a/Configure b/Configure
index b7d669c..f1076b9 100755
--- a/Configure
+++ b/Configure
@@ -267,7 +267,7 @@ if ($config{shlib_version_history} ne "") { $config{shlib_version_history} .= ":
 ($config{major}, $config{minor})
     = ($config{version} =~ /^([0-9]+)\.([0-9\.]+)/);
 ($config{shlib_major}, $config{shlib_minor})
-    = ($config{shlib_version_number} =~ /^([0-9]+)\.([0-9\.]+)/);
+    = ($config{shlib_version_number} =~ /^([0-9]+)\.([0-9\.]+[a-z]?)/);
 die "erroneous version information in opensslv.h: ",
     "$config{major}, $config{minor}, $config{shlib_major}, $config{shlib_minor}\n"
     if ($config{major} eq "" || $config{minor} eq ""
```

**Comment by levitte** [comment]
No. 

Actually, the edit number shouldn't be part of it either, it really should just be .so.1.0, nothing more.

**Comment by dvarrazzo** [comment]
@levitte that's the choice of what `SHLIB_VERSION_NUMBER` should be, and I agree if you want to only report binary compatibility then `1.0` is fine.

But because someone wants to explicit in the `.so` name the full version, I think if they go to the length of changing `SHLIB_VERSION_NUMBER`, the shlib name should be consistent with it: this is the point of this ticket.

If `.so.1.0` is fine the above regexp should be:

```diff
diff --git a/Configure b/Configure
index b7d669c..f1076b9 100755
--- a/Configure
+++ b/Configure
@@ -267,7 +267,7 @@ if ($config{shlib_version_history} ne "") { $config{shlib_version_history} .= ":
 ($config{major}, $config{minor})
     = ($config{version} =~ /^([0-9]+)\.([0-9\.]+)/);
 ($config{shlib_major}, $config{shlib_minor})
-    = ($config{shlib_version_number} =~ /^([0-9]+)\.([0-9\.]+)/);
+    = ($config{shlib_version_number} =~ /^([0-9]+)\.([0-9]+)/);
 die "erroneous version information in opensslv.h: ",
     "$config{major}, $config{minor}, $config{shlib_major}, $config{shlib_minor}\n"
     if ($config{major} eq "" || $config{minor} eq ""
```

as it is now it seems to me it does just half the job.

**Comment by levitte** [comment]
Well, if *you* want to stop down the SONAME version to whatever you like, that's your choice of course.  I'll have you note, btw, that for all 1.0.*x* versions, `SHLIB_VERSION_NUMBER` has stayed the same, `1.0.0`.  What it is exactly doesn't really matter that much, what matters is that *it stays the same as long as the shared library stays upwardly compatible*.  That last bit has me sharply question your desire to have the shared library version be the exact same as the OpenSSL version.  By doing so, you're thwarting any dynamic upgradability of the library.

**Comment by dvarrazzo** [comment]
The use case is bundling a Python library together with the dynamic libraries it uses in a "wheel" package (https://github.com/pypa/manylinux). By displaying the full package version in the lib name users can use `ls` to know the version and assess the need of an upgrade, instead of using `string | grep 'OpenSSH '` or exposing `SSLeay_version()` somehow.

Dynamic upgrade is not a requirement for these packages, and not even possible because the sonames are mangled by the `manylinux` packaging chain which adds the file hash to the soname, resulting e.g. in `libssl-08959156.so.1.0.2k` (people can still build the package from source if they want dynamic upgrades: mine is a convenience to allow installing the package on clients without dev tools).

**Comment by levitte** [comment]
Ok...

First of all, like I already mentioned, your patch will not change anything, because the value of the version number is taken from `opensslv.h` (`crypto/opensslv.h` for 1.0.2, `include/openssl/opensslv.h` for 1.1.0 and on).  In OpenSSL 1.0.*x*, this is the line that's used:

``` C
# define SHLIB_VERSION_NUMBER "1.0.0"
```

In OpenSSL 1.1.*x*, it looks like this:

``` C
# define SHLIB_VERSION_NUMBER "1.1"
```

Now, if I understand you correctly, you have a packager's role and would like to do the same kind of mangling other packagers are doing.  There are all kinds of methods to do so, most people modify the `Makefile` that `Configure created`.  In your case, though, it seems the solution is even easier.  Something like this (should work for both OpenSSL 1.0.2 and 1.1.0):

```
shlibvmajor=`grep ^VERSION= Makefile | cut -f2 -d= | cut -f1 -d.`
shlibvminor=`grep ^VERSION= Makefile | cut -f2 -d= | cut -f2- -d.`
make SHLIB_MAJOR=$shlibvmajor SHLIB_MINOR=$shlibvminor SHLIB_VERSION_NUMBER=$shlibvmajor.$shlibvminor
```

I just tested that for OpenSSL 1.0.2l-dev, and got this:

```
$ ls -l lib*
-rw-r--r-- 1 levitte levitte 4539610 mar  9 13:12 libcrypto.a
-rw-r--r-- 1 levitte levitte     297 mar  9 13:12 libcrypto.pc
lrwxrwxrwx 1 levitte levitte      23 mar  9 13:12 libcrypto.so -> libcrypto.so.1.0.2l-dev
-rwxr-xr-x 1 levitte levitte 2703792 mar  9 13:12 libcrypto.so.1.0.2l-dev
-rw-r--r-- 1 levitte levitte  768744 mar  9 13:12 libssl.a
-rw-r--r-- 1 levitte levitte     298 mar  9 13:12 libssl.pc
lrwxrwxrwx 1 levitte levitte      20 mar  9 13:12 libssl.so -> libssl.so.1.0.2l-dev
-rwxr-xr-x 1 levitte levitte  517160 mar  9 13:12 libssl.so.1.0.2l-dev
$ ./util/shlib_wrap.sh ldd apps/openssl
	linux-vdso.so.1 (0x00007ffe5cd66000)
	./util/../libcrypto.so.1.0.2l-dev (0x00007f02cacdb000)
	./util/../libssl.so.1.0.2l-dev (0x00007f02caa69000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f02ca81b000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f02ca47d000)
	/lib64/ld-linux-x86-64.so.2 (0x0000562fec05e000)
```

**Comment by dvarrazzo** [comment]
Sorry if my patch is wrong: I effectively implemented it on openssl 1.0.2, haven't tested it for 1.1. In my build chain I'm already [sed'ing around](https://github.com/psycopg/psycopg2-wheels/blob/f6fd148/scripts/build_libpq.sh#L21-L27) to obtain the desired behaviour. I thought it was something useful out-of-the-box: that's your call.

**Comment by levitte** [comment]
Not interesting at all for us.  However, the `make` line in my comment above might be of interest to you, as it would take away *all* your need to "sed around".  Your call, though.

I think we're done here at this point.
